File: agent/security-agent.js
#!/usr/bin/env node
import { Kafka } from 'kafkajs';

import {
  createObservabilityContext,
  emitEvent,
} from '../src/lib/observability.js';
import { createIncident, shouldRaiseIncident } from '../src/lib/security-events.js';

const BOOTSTRAP = process.env.KAFKA_BOOTSTRAP_SERVERS || 'kafka:9092';
const INPUT_TOPIC = process.env.KAFKA_INPUT_TOPIC || 'security.dependency.events';
const INCIDENT_TOPIC = process.env.KAFKA_INCIDENT_TOPIC || 'security.incidents';
const GROUP_ID = process.env.KAFKA_GROUP_ID || 'dependency-security-agent';

const obs = createObservabilityContext({ tool: 'security_agent' });

async function main() {
  const kafka = new Kafka({ brokers: BOOTSTRAP.split(',').map((item) => item.trim()).filter(Boolean) });
  const consumer = kafka.consumer({ groupId: GROUP_ID });
  const producer = kafka.producer();

  const shutdown = async () => {
    await emitEvent(obs, 'agent_shutdown', {});
    try {
      await consumer.disconnect();
    } catch {
      // Ignore shutdown errors
    }
    try {
      await producer.disconnect();
    } catch {
      // Ignore shutdown errors
    }
    process.exit(0);
  };

  process.once('SIGINT', () => {
    void shutdown();
  });
  process.once('SIGTERM', () => {
    void shutdown();
  });

  await producer.connect();
  await consumer.connect();
  await consumer.subscribe({ topic: INPUT_TOPIC, fromBeginning: true });

  await emitEvent(obs, 'agent_started', {
    input_topic: INPUT_TOPIC,
    incident_topic: INCIDENT_TOPIC,
  });
  process.stdout.write(`Listening on topic: ${INPUT_TOPIC}\n`);

  await consumer.run({
    eachMessage: async ({ message }) => {
      try {
        const event = JSON.parse(String(message.value || '{}'));
        const status = event?.status;

        if (shouldRaiseIncident(status)) {
          const incident = createIncident(event);
          await producer.send({ topic: INCIDENT_TOPIC, messages: [{ value: JSON.stringify(incident) }] });
          await emitEvent(obs, 'incident_published', {
            repository: event?.repository,
            status,
            incident_topic: INCIDENT_TOPIC,
          });
          process.stdout.write(`Incident created for repo=${event?.repository} status=${status}\n`);
        } else {
          await emitEvent(obs, 'event_allowed', {
            repository: event?.repository,
            status,
          }, 'debug');
          process.stdout.write(`Allowed dependency set for repo=${event?.repository}\n`);
        }
      } catch (error) {
        await emitEvent(obs, 'agent_processing_error', {
          error: String(error?.message || error),
        }, 'error');
        process.stderr.write(`Processing error: ${error?.message || error}\n`);
      }
    },
  });
}

try {
  await main();
} catch (error) {
  await emitEvent(obs, 'agent_startup_error', {
    error: String(error?.message || error),
  }, 'error');
  process.stderr.write(`Agent startup failure: ${error?.message || error}\n`);
  process.exit(1);
}

File: agent/security_agent.py
#!/usr/bin/env python3
import json
import os
import sys
from datetime import datetime, timezone

from confluent_kafka import Consumer, Producer

BOOTSTRAP = os.environ.get("KAFKA_BOOTSTRAP_SERVERS", "kafka:9092")
INPUT_TOPIC = os.environ.get("KAFKA_INPUT_TOPIC", "security.dependency.events")
INCIDENT_TOPIC = os.environ.get("KAFKA_INCIDENT_TOPIC", "security.incidents")
GROUP_ID = os.environ.get("KAFKA_GROUP_ID", "dependency-security-agent")


def utc_now() -> str:
    return datetime.now(timezone.utc).isoformat()


def create_incident(event: dict) -> dict:
    severity = "critical" if event.get("status") == "block" else "high"

    return {
        "incident_type": "dependency_supply_chain_risk",
        "created_at": utc_now(),
        "severity": severity,
        "repository": event.get("repository"),
        "pipeline_id": event.get("pipeline_id"),
        "commit_sha": event.get("commit_sha"),
        "status": event.get("status"),
        "summary": event.get("summary", {}),
        "blocked": event.get("blocked", []),
        "quarantined": event.get("quarantined", []),
        "recommended_actions": [
            "Fail pipeline immediately" if event.get("status") == "block" else "Quarantine artifact",
            "Open security incident",
            "Require manual security review",
            "Prevent production promotion",
            "Update central denylist/policy",
        ],
    }


def main() -> int:
    consumer = Consumer(
        {
            "bootstrap.servers": BOOTSTRAP,
            "group.id": GROUP_ID,
            "auto.offset.reset": "earliest",
        }
    )
    producer = Producer({"bootstrap.servers": BOOTSTRAP})

    consumer.subscribe([INPUT_TOPIC])
    print(f"Listening on topic: {INPUT_TOPIC}")

    try:
        while True:
            msg = consumer.poll(1.0)
            if msg is None:
                continue
            if msg.error():
                print(f"Consumer error: {msg.error()}", file=sys.stderr)
                continue

            try:
                event = json.loads(msg.value().decode("utf-8"))
                status = event.get("status")

                if status in ("block", "quarantine"):
                    incident = create_incident(event)
                    producer.produce(INCIDENT_TOPIC, json.dumps(incident).encode("utf-8"))
                    producer.flush()
                    print(f"Incident created for repo={event.get('repository')} status={status}")
                else:
                    print(f"Allowed dependency set for repo={event.get('repository')}")
            except Exception as exc:
                print(f"Processing error: {exc}", file=sys.stderr)

    except KeyboardInterrupt:
        pass
    finally:
        consumer.close()

    return 0


if __name__ == "__main__":
    sys.exit(main())
